flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java

```diff
@@ -0,0 +1,185 @@
+import org.apache.flink.runtime.state.StateHandleID;
+import org.apache.flink.runtime.state.StreamStateHandle;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Data transfer utils for RocksDbKeyedStateBackend.
+ */
+public class RocksDbStateDataTransfer {
+
+	public static void transferAllStateDataToDirectory(
+		IncrementalKeyedStateHandle restoreStateHandle,
+		Path dest,
+		int restoredThreadNum,
+		CloseableRegistry closeableRegistry) throws Exception {
+
+		final Map<StateHandleID, StreamStateHandle> sstFiles =
+			restoreStateHandle.getSharedState();
+		final Map<StateHandleID, StreamStateHandle> miscFiles =
+			restoreStateHandle.getPrivateState();
+
+		transferAllDataFromStateHandles(sstFiles, dest, restoredThreadNum, closeableRegistry);
+		transferAllDataFromStateHandles(miscFiles, dest, restoredThreadNum, closeableRegistry);
+	}
+
+	/**
+	 * Copies all the files from the given stream state handles to the given path, renaming the files w.r.t. their
+	 * {@link StateHandleID}.
+	 */
+	private static void transferAllDataFromStateHandles(
+		Map<StateHandleID, StreamStateHandle> stateHandleMap,
+		Path restoreInstancePath,
+		int restoredThreadNum,
+		CloseableRegistry closeableRegistry
+		) throws Exception {
+
+		AtomicBoolean running = new AtomicBoolean(true);
+		List<Runnable> runnables = getRunnablesForStateHandles(stateHandleMap, restoreInstancePath, closeableRegistry, running);
+
+		if (enableMultiThreadDownload(restoredThreadNum)) {
+			runRunnalbesMultithread(runnables, restoredThreadNum, closeableRegistry, running);
+		} else {
+			runRunnablesSingleThread(runnables);
+		}
+	}
+
+	private static List<Runnable> getRunnablesForStateHandles(
+		Map<StateHandleID, StreamStateHandle> stateHandleMap,
+		Path restoreInstancePath,
+		CloseableRegistry closeableRegistry,
+		AtomicBoolean running
+		) {
+		List<Runnable> runnables = new ArrayList<>(stateHandleMap.size());
+		for (Map.Entry<StateHandleID, StreamStateHandle> entry : stateHandleMap.entrySet()) {
+			StateHandleID stateHandleID = entry.getKey();
+			StreamStateHandle remoteFileHandle = entry.getValue();
+
+			Path path = new Path(restoreInstancePath, stateHandleID.toString());
+			runnables.add(() -> copyStateDataHandleData(path, remoteFileHandle, closeableRegistry, running));
+		}
+		return runnables;
+	}
+
+	private static void runRunnalbesMultithread(
+		List<Runnable> runnables,
+		int restoredThreadNum,
+		CloseableRegistry closeableRegistry,
+		AtomicBoolean running) throws ExecutionException, InterruptedException, IOException {
+
+		final ExecutorService executorService = enableMultiThreadDownload(restoredThreadNum) ?
```

Comment:
Here we can already assume that `enableMultiThreadDownload  == true`, right? So it can be just:
`final ExecutorService executorService = Executors.newFixedThreadPool(restoredThreadNum);`

I also think it can be just `threadNum` here because it can be used also for upload same way.